### PR TITLE
fix: add retry for initial trivy scan to handle 429's from ghcr on DB update

### PR DESCRIPTION
**Pull request by cameronmeissner**
… update

**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:

```
none
```

### CODE REVIEW

**Review by jason1028kr — APPROVED** [review]

**Comment by ganeshkumarashok** [comment]
Can we consider mirroring it from ghcr to mcr? Is there one that already exists? That could help us avoid throttling.

File: vhdbuilder/packer/trivy-scan.sh [hunk 1]
```diff
@@ -114,7 +128,9 @@ chmod a+x ${MODULE_NAME}
 # shellcheck disable=SC2155
 export PATH="$(pwd):$PATH"
 
-./trivy --scanners vuln rootfs -f json --skip-dirs /var/lib/containerd --ignore-unfixed --severity ${SEVERITY} -o "${TRIVY_REPORT_ROOTFS_JSON_PATH}" /
+# we do a delayed retry here since it's possible we'll get rate-limited by ghcr.io, which hosts the vulnerability DB
+retrycmd_if_failure 10 60 600 ./trivy --scanners vuln rootfs -f json --skip-dirs /var/lib/containerd --ignore-unfixed --severity ${SEVERITY} -o "${TRIVY_REPORT_ROOTFS_JSON_PATH}" /
```
Comment (ganeshkumarashok) [inline vhdbuilder/packer/trivy-scan.sh:132]: 6600 seconds/110 minutes is the max timeout time for: retrycmd_if_failure 10 60 600.  That's a long time, more than certain VHD builds. I think we need to 1) reduce the timeout time, 2) (more medium-term) consider preventing a trivy scan throttling issue from blocking the full build.

**Comment by cameronmeissner** [comment]
`consider using MS Defender tooling instead`

would this be an alternative to trivy altogether? or just another data source we can point trivy towards?

File: vhdbuilder/packer/trivy-scan.sh (hunk 1 above)
Comment (cameronmeissner) [reply vhdbuilder/packer/trivy-scan.sh:132]: I highly doubt we'd ever run into trivy stages taking that long (if all 10 retries failed on 429 then it'd be a little over 10 minutes until we see the failure in the worst case with the current config)

though overall I agree this is a bandaid and not a solution

Comment (ganeshkumarashok) [reply vhdbuilder/packer/trivy-scan.sh]: I've had to fix such errors for container image pull retries - it's quite time consuming to wait till failure (even for a different reason). And the ADO build logs don't show up until it completes, so it's hard to know why it failed.

**Comment by ganeshkumarashok** [comment]
> `consider using MS Defender tooling instead`
> 
> would this be an alternative to trivy altogether? or just another data source we can point trivy towards?

if we go this route, it'll be an alternate to trivy altogether

File: vhdbuilder/packer/trivy-scan.sh (hunk 1 above)
Comment (cameronmeissner) [reply vhdbuilder/packer/trivy-scan.sh]: reduced retry wait time to 30s from 60s

**Pull request merged**